src/io_uring/src/lib.rs

```diff
@@ -234,3 +234,233 @@ impl IoUring {
+    use utils::tempfile::TempFile;
+    use vm_memory::{Bytes, MmapRegion, VolatileMemory};
+
+    fn drain_cqueue(ring: &mut IoUring) {
+        while let Some(entry) = ring.pop::<usize>().unwrap() {
+            assert!(entry.result().is_ok());
+
+            // Assert that there were no partial writes.
+            let count = entry.result().unwrap();
+            let user_data = entry.user_data() as u32;
+            assert_eq!(count, user_data);
+        }
+    }
+
+    fn setup_mem_region(len: usize) -> MmapRegion {
+        const PROT: i32 = libc::PROT_READ | libc::PROT_WRITE;
+        const FLAGS: i32 = libc::MAP_ANONYMOUS | libc::MAP_PRIVATE;
+
+        let ptr = unsafe { libc::mmap(std::ptr::null_mut(), len, PROT, FLAGS, -1, 0) };
+
+        if (ptr as isize) < 0 {
+            panic!("Mmap failed with {}", std::io::Error::last_os_error());
+        }
+
+        unsafe {
+            // Use the raw version because we want to unmap memory ourselves.
+            MmapRegion::build_raw(ptr as *mut u8, len, PROT, FLAGS).unwrap()
+        }
+    }
+
+    fn free_mem_region(region: MmapRegion) {
+        unsafe { libc::munmap(region.as_ptr() as *mut libc::c_void, region.len()) };
+    }
+
+    fn read_entire_mem_region(region: &MmapRegion) -> Vec<u8> {
+        let mut result = vec![0u8; region.len()];
+        let count = region.as_volatile_slice().read(&mut result[..], 0).unwrap();
+        assert_eq!(count, region.len());
+        result
+    }
+
+    fn arbitrary_rw_operation(file_len: u32) -> impl Strategy<Value = Operation<u32>> {
+        (
+            // OpCode: 0 -> Write, 1 -> Read.
+            0..2,
+            // Length of the operation.
+            0u32..file_len,
+        )
+            .prop_flat_map(move |(op, len)| {
+                (
+                    // op
+                    Just(op),
+                    // len
+                    Just(len),
+                    // offset
+                    (0u32..(file_len - len)),
+                    // mem region offset
+                    (0u32..(file_len - len)),
+                )
+            })
+            .prop_map(move |(op, len, off, mem_off)| {
+                // We actually use an offset instead of an address, because we later need to modify
+                // the memory region on which the operation is performed, based on the opcode.
+                let mut operation = match op {
+                    0 => Operation::write(0, mem_off as usize, len, off.into(), len),
+                    _ => Operation::read(0, mem_off as usize, len, off.into(), len),
+                };
+
+                // Make sure the operations are executed in-order, so that they are equivalent to
+                // their sync counterparts.
+                operation.set_linked();
+                operation
+            })
+    }
+
+    #[test]
+    fn proptest_read_write_correctness() {
+        if KernelVersion::get().unwrap() < KernelVersion::new(5, 10, 0) {
+            return;
```

Comment:
Can we use the `skip_if_kernel_lt_5_10!();` macro here ?

Reply:
done. moved the macro in `utils`